050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,821 @@
+So a node which moves will start from _zero_ with empty disk and will need to re-sync all the data.
+
+### Matching the node pools to the Kafka cluster
+
+A particular `KafkaNodePool` resource can belong only to a single Kafka cluster.
+The node pools will use the same mechanism to define to which cluster they belong as we use today for `KafkaConnector`, `KafkaTopic` or `KafkaUSer` resources.
+`strimzi.io/cluster` label will be set on the `KafkaNodePool` and its value will be the name of the `Kafka` custom resource to which it belongs.
+For example, the label `strimzi.io/cluster: my-cluster` will mean that this node pool belongs to a Kafka cluster named `my-cluster`.
+The `Kafka` and `KafkaNodePool` resources always have to be in the same namespace.
+
+#### Example
+
+The following example shows how the `Kafka` and `KafkaNodePool` resources might combine to setup a Kafka cluster with 6 different brokers:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  namespace: my-namespace
+spec:
+  kafka:
+    version: 3.4.0
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+  # ...
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: big-nodes
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 64Gi
+      cpu: 8
+    limits:
+      memory: 64Gi
+      cpu: 12
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+    - id: 1
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: small-nodes
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 4Gi
+      cpu: 1
+    limits:
+      memory: 4Gi
+      cpu: 2
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 100Gi
+      class: regular-storage
+```
+
+#### Protecting `KafkaNodePool` resources from being shared between multiple Kafka clusters
+
+The `strimzi.io/cluster` label should guarantee that the node pool will match only one Kafka cluster.
+Additional protection will be made against undesired moves between different Kafka clusters when the `strimzi.io/cluster` label is modified.)
+The `.status` section of the `KafkaNodePool` resource will contain the Kafka cluster ID of the cluster it belongs to.
```

Comment:
I think also having the node pools reflected in the `Kafka` status can be very handy to have a single view of the cluster. I can think also if it would reflect the ignored fields from the nodepools that can help users troubleshoot undesired Kafka setups. It can be limited to the case where there's a conflict to reflect the effective value.

Reply:
What exactly would you want to have reflected there? I'm open to ideas. But keep in mind that there are size limitations to what we can put there both from the CRD schema perspective as well as from the resource size limits.